keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+---
+title: Object Bucket Provisioning
+authors:
+  - "@jeffvance"
+  - "@copejon"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-storage
+reviewers:
+  - "@saad-ali"
+  - "@alarge"
+  - "@erinboyd"
+  - "@guymguym"
+  - "@travisn"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-25
+last-updated: 2019-11-15
+status: implementable
+see-also:
+  - "https://github.com/kube-object-storage/lib-bucket-provisioner"
+  - "https://github.com/kube-object-storage/lib-bucket-provisioner/blob/master/doc/design/object-bucket-lib.md"
+  - "https://github.com/kube-object-storage/lib-bucket-provisioner/tree/master/doc/examples"
+---
+
+# Object Bucket Provisioning
+
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Custom Resources](#custom-resources)
+  - [Moving Parts](#moving-parts)
+  - [Interfaces](#interfaces)
+  - [Library Usage](#library-usage)
+  - [Current Restrictions](#current-restrictions)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Examples](#examples)
+      - [Alpha -&gt; Beta Graduation](#alpha---beta)
+      - [Beta -&gt; GA Graduation](#beta---ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+<!-- /toc -->
+
+## Summary
+
+Kubernetes natively supports dynamic provisioning for file and block storage but lacks support for object bucket provisioning.
+Without a formal API to describe and manage bucket provisioning, storage providers are left to write their own provisioners including all of the necessary controller code.
+This "wild west" landscape for bucket provisioning is potentially confusing, overly complex, inconsistent, may slow down Kubernetes adoption by object store users, and may hinder application portability.
+
+We propose a common Kubernetes control-plane API for the management of object store bucket lifecycles which has no dependencies on the underlying providers.
+
+## Motivation
+
+By defining a Kubernetes API for object bucket provsioning, we provide a consistent framework for all object store providers and thus enhance the end user's storage experience. This API allows object store vendors to focus on how they wish to provision buckets instead of writing and testing the controller aspects of provisioning.
+
+### Goals
+
++ Define a _control-plane_ object bucket management API thus relieving object store providers from Kubernetes controller details.
++ Minimize technical ramp-up for storage vendors who have already contributed external storage provisioners (file and block).
```

Comment:
I don't understand why "who have already contributed..." is relevant.  Minimizing ramp-up is good, but again (per previous comment), I don't think there is any inherent association between object and volumes (block and file).

Reply:
Agree there is not necessarily an inherent association, per your comment above. Some storage vendor provide file, block and object store. My thinking was that there _could_ be some leverage if a file/block vendor decided to write their own bucket provisioner.

Reply:
>  if a file/block vendor decided to write their own bucket provisioner.

The feedback we've gotten on the bucket provisioning library ("why not make it CSI-like?") makes even more sense when considering adoption by vendors into existing provisioners.  For one, CSI being the standard model for file/block means that it may actually be more cumbersome to incorporate the library.  Being written in go immediately creates more work for provisioner authors who don't use the language.  And the library pattern be in fact be unfamiliar to provisioner authors who jumped straight to CSI.